api/agent/agent.go

```diff
@@ -718,49 +737,49 @@ func (a *agent) runHot(ctx context.Context, call *call, tok ResourceToken, state
 			}
 
 			isFrozen := false
-			elapsed := time.Duration(0)
-			freezerTicker := a.freezeIdleMsecs
-			idleTimeout := time.Duration(call.IdleTimeout) * time.Second
 
 			done := make(chan struct{})
 			state.UpdateState(ctx, ContainerStateIdle, call.slots)
 			s := call.slots.queueSlot(&hotSlot{done, errC, container, nil})
 
+			stopTimer(freezeTimer)
+			stopTimer(idleTimer)
+			stopTimer(ejectTimer)
+
+			freezeTimer.Reset(a.freezeIdleMsecs)
+			idleTimer.Reset(time.Duration(call.IdleTimeout) * time.Second)
+			ejectTimer.Reset(a.ejectIdleMsecs)
+
 			for {
 				select {
 				case <-s.trigger: // slot already consumed
 				case <-ctx.Done(): // container shutdown
 				case <-a.shutdown: // server shutdown
-				case <-time.After(idleTimeout): // in case idleTimeout < a.freezeIdleMsecs or idleTimeout < a.ejectIdleMsecs
-				case <-time.After(freezerTicker):
-					elapsed += a.freezeIdleMsecs
-
-					freezerTicker = math.MaxInt64 // do not fire again
-
-					if elapsed < idleTimeout { // in case idleTimeout <= a.freezeIdleMsecs
-						if !isFrozen {
-							err := cookie.Freeze(ctx)
-							if err != nil {
-								logger.WithError(err).Error("freeze error")
-								return
-							}
-							isFrozen = true
+				case <-idleTimer.C:
+				case <-freezeTimer.C:
+					if !isFrozen {
+						err := cookie.Freeze(ctx)
+						if err != nil {
+							logger.WithError(err).Error("freeze error")
+							return
```

Comment:
only issue is I think since we took the freezeTimer.C cookie that the deferred stop of that timer will wait for the cookie since it didn't stop the timer, but nobody will ever send a cookie (we'll get blocked in defer forever and leak this goroutine)